runtime/src/leader_bank_notifier.rs

```diff
@@ -0,0 +1,161 @@
+use {
+    crate::bank::Bank,
+    solana_sdk::slot_history::Slot,
+    std::{
+        sync::{Arc, Condvar, Mutex, Weak},
+        time::{Duration, Instant},
+    },
+};
+
+/// Tracks leader status of the validator node and notifies when:
+///     1. A leader bank initiates (=PoH-initiated)
+///     2. A leader slot completes (=PoH-completed)
+#[derive(Debug, Default)]
+pub struct LeaderBankNotifier {
+    /// Current state (slot, bank, and status) of the system
+    state: Mutex<SlotAndBankWithStatus>,
+    /// CondVar to notify status changes and waiting
+    condvar: Condvar,
+}
+
+/// Leader status state machine for the validator.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+enum Status {
+    /// The leader bank is not currently available. Either not initialized, or PoH-completed bank.
+    #[default]
+    StandBy,
+    /// PoH-initiated bank is available.
+    InProgress,
+}
+
+#[derive(Debug, Default)]
+struct SlotAndBankWithStatus {
+    status: Status,
+    slot: Slot,
```

Comment:
nit: the `Some(0)` case in https://github.com/solana-labs/solana/pull/30395/files#r1144267697 was possible because of this not `Option<Slot>`. not all of solana code aren't diligent to distinguish block 0 (https://explorer.solana.com/block/0) and None. but I'd prefer Option here a bit.

Reply:
at least, `Option<Slot>`  can remind us of thinking twice (or thrice, lol) for the uninitialized case silently introduced `#[derive(Default)]`...

Reply:
100% => 9f41c41c5b516b981c5127ea9553e52f12fe6f88. I even had errors in my tests from thinking the default_for_tests bank was slot 1 not slot 0 🤦